infrastructure/src/main/java/org/corfudb/infrastructure/BaseServer.java

```diff
@@ -125,6 +188,40 @@ public synchronized void handleMessageSetEpoch(@NonNull CorfuPayloadMsg<Long> ms
     }
 
     /**
+     * Respond to a epoch change request.
+     * This method also executes sealing logic on each individual server type.
+     * @param req The incoming request message.
+     * @param ctx The channel context.
+     * @param r The server router.
+     */
+    @RequestHandler(type = RequestPayloadMsg.PayloadCase.SEAL_REQUEST)
+    private synchronized void handleSeal(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        try {
+            final long epoch = req.getPayload().getSealRequest().getEpoch();
+            String remoteHostAddress;
+            try {
+                remoteHostAddress = ((InetSocketAddress)ctx.channel().remoteAddress()).getAddress().getHostAddress();
+            } catch(NullPointerException ex) {
+                remoteHostAddress = "unavailable";
+            }
+
+            log.info("handleSeal[{}]: Received SEAL from (clientId={}:{}), moving to new epoch {},",
+                    req.getHeader().getRequestId(), req.getHeader().getClientId(), remoteHostAddress, epoch);
+
+            serverContext.setServerEpoch(epoch, r);
+            HeaderMsg responseHeader = getHeaderMsg(req.getHeader(), false, true);
+            ResponseMsg response = getResponseMsg(responseHeader, getSealResponseMsg());
+            r.sendResponse(response, ctx);
+        } catch (WrongEpochException e) {
+            log.debug("handleSeal[{}]: Rejected SEAL current={}, requested={}", req.getHeader().getRequestId(),
+                    e.getCorrectEpoch(), req.getPayload().getSealRequest().getEpoch());
+
+            r.sendWrongEpochError(req.getHeader(), ctx, e.getCorrectEpoch());
```

Comment:
Good point. We did this a bit differently than the others because the other servers need to send a WrongEpochError too. This way, all of the server RPCs simply need to invoke this method with the original request header as well as the correct epoch.

Reply:
Note: this has since been addressed and is handled like the other responses.